### PR TITLE
[Spike] Evaluate blob storage copy

### PR DESCRIPTION
**Pull request by ouphi**
In this spike we evaluated possible ways to copy blobs without downloading the data.

Evaluated topics:
- Copy with Java SDK
- Copy using AzCopy
- Copy vs streaming in EDC
- Observability of copy operations
- Integration testing approach

The PR contains sample code and documentation.

Spike done by @Izzzu and @ouphi

### CODE REVIEW

File: blobspike/README.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
+# Azure Blob Storage Copy
+
+We evaluated the possibility to copy a blob.
+
+## Copy with Java SDK
+
+We evaluated the [Java azure storage client library](https://docs.microsoft.com/en-us/java/api/overview/azure/storage-blob-readme?view=azure-java-stable) to copy blobs.
```
Comment (marcgs) [inline blobspike/README.md:7]: Remove language from link paths so that the browser language is used

File: blobspike/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+/*
+ *  Copyright (c) 2021 Bayerische Motoren Werke Aktiengesellschaft (BMW AG)
```
Comment (marcgs) [inline blobspike/build.gradle.kts:2]: BMW?

File: blobspike/README.md [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+We have 2 solutions to transfer data from a blob to another blob.
+
+- Use the [BlobClient](https://docs.microsoft.com/en-us/java/api/com.azure.storage.blob.blobclient?view=azure-java-stable) beginCopy to trigger a copy of source blob to destination blob.
+- Use the [BlockBlobClient](https://docs.microsoft.com/en-us/java/api/com.azure.storage.blob.specialized.blockblobclient?view=azure-java-stable) to get a stream from blob to VM and transfer from VM to blob with another stream.
+
+Pros and cons to use beginCopy in EDC:
+- Pros: During the copy, Network I/O is only between the 2 storage accounts. The client does not need to download any data.
+- Cons: The source storage account needs to be accessible from the internet. With the stream, the provider and the source storage account could be in the same private network. Then, only the destination storage account would need to be accessible from the internet.
+
+### Azure Blob Copy within same container
```
Comment (marcgs) [inline blobspike/README.md:35]: Is this related to "Evaluation of copy vs streaming"? Why is it a subsection of it?

File: blobspike/README.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
+# Azure Blob Storage Copy
```
Comment (marcgs) [inline blobspike/README.md:1]: I suggest revisiting the structure of the document. I have issues following which sections are related to one another. Perhaps it would be good to provide a table summarizing options with duration, pro and cons?

File: blobspike/README.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+We evaluated the possibility to copy a blob.
```
Comment (cpeeyush) [inline blobspike/README.md:3]: may be we can say that we evaluated server side copy ?

File: blobspike/build.gradle.kts (hunk 2 above)
Comment (Izzzu) [reply blobspike/build.gradle.kts]: copy paste mistake, fixed

File: blobspike/README.md [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+We evaluated the [Java azure storage client library](https://docs.microsoft.com/en-us/java/api/overview/azure/storage-blob-readme?view=azure-java-stable) to copy blobs.
+
+### Evaluate if copy can be done from source to destination, with the client not handling any data flow
+
+The Java Azure Storage Blob client library is on top of the [Blob service REST API](https://docs.microsoft.com/en-us/rest/api/storageservices/blob-service-rest-api).
+The client calls [an endpoint](https://docs.microsoft.com/en-us/rest/api/storageservices/copy-blob-from-url) to trigger the copy of a blob.
+The client receives an answer, the copy is then done asynchronously.
+In order to know if the copy is finished, the client needs to [get the blob properties](https://docs.microsoft.com/en-us/rest/api/storageservices/get-blob-properties) by calling the REST API in order to know if the copy operation is finished.
+
+### Copy duration
```
[the remark this answers is not in the document]
Comment (ouphi) [reply blobspike/README.md:16]: We though that it was better to get indication about the durations before digging into details of the implementation.

**Comment by cpeeyush** [comment]
Closing as evaluation completed with team.
